### ISSUE TITLE
App crashes without calling the errorhandler

### ISSUE THREAD

**Issue by ecosSystem** [X-Sharp/XSharpPublic#881]
Sometimes crashes don't call the errorhandler and the app just dies without showing the error-dialog.
I could reproduce it in the standard VOMDIApp.

In the Fileopen-Method of Standardshellwindow I implemented a wanted crash. If you call the method by selecting the Menuitem File / Open the errorhandler is called. If you call the method via shortcut (CTRL-O) the app dies without message.
The code that causes the crash:

CLASS StandardShellWindow INHERIT ShellWindow

METHOD FileOpen() 
//LOCAL oOD AS OpenDialog
LOCAL oOD AS SaveasDialog

//oOD := OpenDialog{SELF, "*.dbf"}
oOD := SELF:oD
oOD:Show()
	
IF !Empty(oOD:FileName)
	SELF:DoOpenFile(oOD:FileName, oOD:ReadOnly)
ENDIF

RETURN SELF	

ACCESS OD
LOCAL oOD AS OpenDialog
oOD := OpenDialog{SELF, "*.dbf"}
RETURN oOD

**Comment by cpyrgas** [comment]
Karl, are you sure this code can cause the error dialog to show? I would be very surprised, because this is an exception that happens in regular code, not in the X# runtime (like dbf access) which is responsible for showing the error dialog.

Such errors can normally be trapped only with TRY...CATCH commands.

**Comment by ecosSystem** [comment]
Chris, I assign an Opendialog to a var defined as SaveasDialog. This definitely causes a crash here...

**Comment by cpyrgas** [comment]
Yes, the crash happens indeed, but that's a crash (an exception thrown) that needs to be caught with TRY..CATCH. Doesn't the debugger in VS stop in the line with the error as you run it? If you are running it outside of VS, then you need to trap the error (this one and every other exception that might happen) probably in the Start() function.

**Comment by cpyrgas** [comment]
Hmmm, maybe I misunderstood. Which dialog are you referring to?

**Comment by ecosSystem** [comment]
My problem is not the crash, I wanted the app to crash. The problem is, that the error-dialog does not always pop up. As I said, invoking the command by clicking the menu-command with the mouse shows the error-dialog, typing CTRL-O does not. In our converted app there seem to be several problems where the app just dies without error-message and this makes it very hard to solve if customers can not report what happened...
I meant this dialog:
![grafik](https://user-images.githubusercontent.com/45791974/144615582-23487f8c-092a-48ac-9d14-d37936732953.png)

**Comment by cpyrgas** [comment]
Karl, this error dialog is displayed when an error happens inside the code of the X# runtime and it is the X# runtime that displays this dialog. When the error happens in your own code, like a division by zero, a wrong type cast, an incorrect index on a collection etc, the X# runtime is not in control, so this dialog is not shown. I know this was different in VO, but in VO the runtime controlled the whole application environment, while in .Net it's the .Net runtime that is in control of things.

In order to trap such errors (not related to the X# runtime), you need to use TRY...CATCH commands, like this:

```
TRY
// code here that could lead to an error
CATCH oException AS Exception
? oException:Message // this objects contains all the information about the error
END TRY
```

of course you can also use a TRY...CATCH also in your Start() function, effectively trapping ALL errors that may happen in your code in one place. But this has the drawback that after trapping the error, the app will be terminated, while if you include also "local" TRY..CATCH commands, then code execution can still continue after the "END" of the TRY..CATCH command.

**Comment by cpyrgas** [comment]
To add a bit more to that, in some cases indeed the X# runtime can still trap the error even if it happens inside your own code. This can happen when there is an "item" of the X# runtime "active" (like a button on a VO window which causes a late bound call to the button's "click" method to execute), so in this case the X# runtime can trap an error that happens in a nested call in response to the button click. 

But if for example the same exception happened in nested code after a button is pressed in a windows forms window and without a late bound call (which is implemented in the X# runtime), then the X# runtime is not in control of the code and cannot show the dialog. 

This is why sometimes you see it and sometimes not, so in order to always be able to catch errors, it's a good idea to install a TRY..CATCH block in the start() function (or Start() method of the app class).

**Comment by ecosSystem** [comment]
Chris, I definitly have a TRY..CATCH in my start function. Please use the attached standard-solution, run the app and click file / open with the mouse. The error message appears, called by the start-function via ErrorDialog(oException). If you restart the app and type CTRL-O no message appears. The app simply dies.
[VOMDIApp1.zip](https://github.com/X-Sharp/XSharpPublic/files/7658586/VOMDIApp1.zip)
..

**Comment by cpyrgas** [comment]
Ah, I hadn't realized that! OK, thanks for the sample, I see the problem now, looking into it...

**Comment by Karl1155** [comment]
Chris, 

I don't know if it helps, but when I install an Unhandled exception handler in the Start(), I see the exception thrown by the [CTRL+O] hotkey

`System.AppDomain.CurrentDomain:UnhandledException += UnhandledExceptionEventHandler{CurrentDomain_UnhandledException}`
	

```
FUNCTION CurrentDomain_UnhandledException( sender AS OBJECT , e AS UnhandledExceptionEventArgs ) AS VOID 

	InfoBox{,,"Unhandled Exception happened !" + crlf + crlf + "The ErrorDialog is opened manually in the next step ..." }:Show()
     
	ErrorDialog ( (Exception) e:ExceptionObject )  
	
RETURN  
```

Karl-Heinz

**Comment by Karl1155** [comment]
In addition:

If you change the "Target" setting from "Windows" to "COnsole" in the app properties, the exception will be displayed in the console window.

**Comment by RobertvanderHulst** [comment]
The difference between selecting the menu option through the shortcut and from the menu is that the code is running through the Win32 API function TranslateAccelerator().
Experimentally I have added my own TranslateAccelerator with the following contents:

```
FUNCTION MyTranslateAccelerator(hwnd AS PTR, hAccTable AS PTR, lpMsg AS _winMSG) AS INT
TRY
    System.AppDomain.CurrentDomain:UnhandledException += UnhandledExceptionEventHandler{CurrentDomain_UnhandledException}
    VAR result := TranslateAccelerator(hwnd, hAccTable, lpMsg)
    System.AppDomain.CurrentDomain:UnhandledException -= UnhandledExceptionEventHandler{CurrentDomain_UnhandledException}
    RETURN result
CATCH e as Exception
    THROW e
END TRY

FUNCTION CurrentDomain_UnhandledException( sender AS OBJECT , e AS UnhandledExceptionEventArgs ) AS VOID
    local ex as Exception
    ex := (Exception) e:ExceptionObject
    IF ex != null .and. ex:InnerException != NULL
        ex := ex:InnerException
    ENDIF

	Eval(ErrorBlock(), Error{ex })
    _Quit()

RETURN
```
And changed App:Exec() to call this accelerator.
Then the error is caught by the UnHandledException handler and can be shown etc.
I suspect that the API funciton TranslateAccelerator plays some dirty tricks itself with unmanaged exception handling.

There is only one place in the GUI SDK where this is called so a solution like this should allow to trap the error.
I checked the System.Windows.Forms classes and they do not call TranslateAccelerator() the way that VO does it.

**Comment by Karl1155** [comment]
Robert,

looks very promising !

i tried your CurrentDomain_UnhandledException() changes:

- the ErrorDialog() window popups.
- the error is written to the VOERROR.LOG.

Karl-Heinz

**Comment by ecosSystem** [comment]
Robert, in VO I have tons of sequences like this:

Begin Sequence
	if ...
		break
	endif
	if ...
		break
	endif
	if ...
		break
	endif

	// here an error might occur
		
	DoSomething

End Sequence

If an error occurred inside the sequence VO threw an error. 
x# seems to treat an error like a break and the app continues after the end sequence.
Of course I can trap this with this recover-statement:

Recover Using oErr
	if oErr <> NIL
		Errorexec(oErr)
	endif
End Sequence

But I really have hundreds (or thousands?) of sequences like this.
Is it possible to throw an error like in VO ?

TIA
Karl

**Comment by RobertvanderHulst** [comment]
Karl.
What do you mean with 'throw an error'. An error that is not going to the Recover Using line?

**Comment by ecosSystem** [comment]
call ErrorExec and call show the error-dialog

**Comment by ecosSystem** [comment]
After thinking a few moments about it, it would really be best to throw an error if no recover using exists. TRY .. END TRY can't be called without CATCH, BEGIN SEQUENCE / END SEQUENCE doesn't need  a RECOVER. Without Recover the app should crash with an error...

**Comment by cpyrgas** [comment]
Karl,

That depends on the type of the error. If the error that happens is for example a dbf related one (so it's happening inside the X# rdd code), then the error dialog will be shown. If it's a program flow error, like a divide by zero, or a wrong type cast etc, then there's no X# runtime active to show the error dialog, so you will not get one. Please also see my related posts above.

About TRY...END TRY, you can actually omit the CATCH (and FINALLY) statements (there was someone I know that insisted on allowing this :)), you will only get a compiler warning if you do that, not an error.

**Comment by ecosSystem** [comment]
Chris, that would make things very difficult for me. Many errors only occur, when customers use the application. If code inside a sequence is not executed and the program just goes on after the end sequence that must lead to serious problems. Nobody notices the problem and nobody can report how things happened. Perhaps it would be an option to let the transporter add a RECOVER with a call to ErrorExec() if none is in the original code (of course as an option). This would help a little...

**Comment by RobertvanderHulst** [comment]
Karl,
If I understand you correctly you are using BEGIN SEQUENCE .. RECOVER.. END SEQUENCE and the BREAK command as a kind of alternative for a GOTO statement.
We have actually considered to implement it like that, however the problem is that the BREAK command may also be called in another function that is called from the function where the SEQUENCE is defined.
The only .Net construct that allows that kind of behavior is the TRY .. CATCH statement.
The BREAK statement is translated into a THROW of a special kind of exception. The RECOVER is translated to a CATCH e AS WrappedException.
What we could do (Chris do you remember why we did not do that) is add a CATCH e as Exception that catches all exceptions and then call the Error handler with this object .

**Comment by ecosSystem** [comment]
Chris, please don't remember, that would save me a lot of work!

**Comment by ecosSystem** [comment]
Robert, yes mainly I use it like a goto in basic. Sometimes (or even very often) execution of code is dependent on many many conditions which would lead to IF IF IF IF IF IF .... ENDIF ENDIF ENDIF ENDIF ENDIF ENDIF. With BEGIN SEQUENCE BREAK END SEQUENCE withoud RECOVER you can avoid this and also avoid returns within functions or methods.

**Comment by RobertvanderHulst** [comment]
Karl,
Are you also BREAKing from code outside of the function where the BEGIN SEQUENCE .. END is declared?

**Comment by ecosSystem** [comment]
never say never, but I think I never do

**Comment by RobertvanderHulst** [comment]
What if we would translate this code
```
FUNCTION Test()
LOCAL e
BEGIN SEQUENCE
   IF SomeCondition
      BREAK 123
   ENDIF
   ? "After IF"
RECOVER USING e
   ? e
END SEQUENCE
```
into this:
```

FUNCTION Test()
LOCAL e
BEGIN SCOPE
   TRY
       // increment counter in runtime so CanBreak() knows that we're inside a BEGIN SEQUENCE
      IF SomeCondition
         e := 123        // Compiler could insert the variable name from the RECOVER statement here
         GOTO BreakLabel 
      ENDIF
      ? "After IF"
      GOTO EndLabel     
      BreakLabel: // compiler generated label name
         ? e
      EndLabel:   // compiler generated label name
    FINALLY
      // decrement the counter in the runtime
    END TRY
END SCOPE
```
However: this would NOT work across functions.

**Comment by ecosSystem** [comment]
Transporter or compiler ?
I'm not sure if this will work. Sometimes I have a recover, sometimes I don't. For me it would be great if missing recovers would be replaced with

RECOVER using oErr
   if oErr <> NIL
       ErrorExec(oErr)
   endif
END SEQUENCE

Maybe as a compiler-switch ?

I could do it globally in VS with replace, but I have problems with oErr, which must be declared as a local var. And I really have more than 1000 places with this.

**Comment by wriedmann** [comment]
PMFJI: I have a lot of similar code, not breaks, but begin/end sequence without recover because the VO runtime executed always the standard errorblock and does not terminated the application. I suspect there are many VO applications using that.

**Comment by cpyrgas** [comment]
I also think that a compiler option to autogenerate a RECOVER block (when there isn't one) that shows an error dialog would be probably helpful.

But what if there's already a RECOVER block in the code? The programmer would need to manually add code to show the error dialog in this case.

Also what cannot work in either case, is allow to continue form the line that the error happened:

BEGIN SEQUENCE
SomeErrorHappenedHere()
ExecuteSomeMoreCode()
END SEQUENCE

in this case, in VO you can choose to ignore the error and keep code executing, so ExecuteSomeMoreCode() can be called. In X#, program control would be directly tranferred to the implicit RECOVER (CATCH) statement. In order to allow "ignore", we would need to make the compiler surround every single line of code with an implicit TRY...CATCH

**Comment by RobertvanderHulst** [comment]
My plan was to let the compiler do that and not allow GOTO as a normal statement. Not just because GOTO invites you to write spaghetti code, but also because GOTO would conflict with the GOTO command that selects a record in the current workarea.

**Comment by ecosSystem** [comment]
Chris, I have places with a recover, but in these cases I usually have changed the Errorblock to {||_break()} and I don't want the error dialog. These are the places where I will use TRY/CATCH in future. "Ignore" is never an option for me. If an error occurs I want to know where and I will correct it.

**Comment by cpyrgas** [comment]
Karl, understood, but we need to implement a solution which hopefully covers as many people as possible!

**Comment by RobertvanderHulst** [comment]
I think we should discuss this during the session this afternoon.
I have an idea how to emulate the VO behavior (even including the retry) but I would have to create some tests to see how that would affect performance.
Basically it would mean that every statement would be wrapped in an action block that is passed to another function that has the error handling and retry mechanism.
Something like this

```
  FUNCTION Start
     LOCAL hFile as PTR
     hFile := FCreate("SomeFile.Txt")
     FWriteLine(hFile, "sometext")
     FClose(hFile)
     RETURN
```
would become
```
  FUNCTION Start
   LOCAL hFile as PTR
     ExecStmt ( { =>
              hFile := FCreate("SomeFile.Txt")
     })
   ExecStmt ( { =>
               FWriteLine(hFile, "sometext")
     })
   ExecStmt ( { =>
              FClose(hFile)
     })
     RETURN

FUNCTION ExecStmt( action AS @@Action) AS VOID
       LOCAL lRepeat := TRUE as LOGIC
       DO WHILE lRepeat
          TRY
              lRepeat := FALSE
              action()
          CATCH e AS Exception
              // Execute ErrorBlock()
              // Set lRepeat based on result of Errorblock
             // Retry sets lRepeat to TRUE
          END TRY
        ENDDO
        RETURN 
```
and in a similar way there could be an ExecExpression, which could be used for expressions in DO WHILE or DO CASE

```
DO WHILE ! Eof()
    ..
ENDDO
```
Would become
```
DO WHILE ExecExpression ( { => ! Eof()} )
    ..
ENDDO


FUNCTION ExecExpression( action AS @@Func<T>) AS T
       LOCAL lRepeat := TRUE as LOGIC
       LOCAL result AS T
       DO WHILE lRepeat
          TRY
              lRepeat := FALSE
              result := Action()
          CATCH e AS Exception
              result := DEFAULT(T)
              // Execute ErrorBlock()
              // Set lRepeat based on result of Errorblock
             // Retry sets lRepeat to TRUE
          END TRY
        ENDDO
        RETURN result
```

I know, this looks like a lot of overhead, but I expect that the performance will still be OK.

**Comment by ecosSystem** [comment]
Robert, I'm not sure about the effect of BEGIN SCOPE / END SCOPE. 
And the Breaklabel should als be implemented with
if e <> NIL
ErrorExec(e)
endif
if there is no original recover-statement. If I have to add the recover statements first it won't help

**Comment by ecosSystem** [comment]
Robert, your latest comment is definitely too high for me, but it really looks like very much overhead. And I think **no** decrease in performance is acceptable for just this issue. Unfortunately I can't take part in the session this afternoon so far, but I'll try to postpone my appointment...

**Comment by RobertvanderHulst** [comment]
> Robert, I'm not sure about the effect of BEGIN SCOPE / END SCOPE.
> And the Breaklabel should als be implemented with
> if e <> NIL
> ErrorExec(e)
> endif
> if there is no original recover-statement. If I have to add the recover statements first it won't help

BEGIN SCOPE / END SCOPE just defines a block.
When you introduce a local variable inside the BEGIN SEQUENCE this is only visible inside the block. BEGIN SCOPE.. does the same.
And yes, I can also generate the code to also call ErrorExec() when needed.

**Comment by RobertvanderHulst** [comment]
> Robert, your latest comment is definitely too high for me, but it really looks like very much overhead. And I think **no** decrease in performance is acceptable for just this issue. Unfortunately I can't take part in the session this afternoon so far, but I'll try to postpone my appointment...

This uses the concept of Lambda expressions / delegates. That is something like a codeblock.
The compiler could wrap every statement or expression in a lambda and pass it to a function that executes the code with error handling in place.

**Comment by RobertvanderHulst** [comment]
> In order to allow "ignore", we would need to make the compiler surround every single line of code with an implicit TRY...CATCH

This is what my example with ExecStmt and ExecExpression demonstrates.
Of course we could also add this in the actual body of the code but that would blow up the size of the code a lot, especially if you also want to allow to retry.

```
FUNCTION Start as VOID
LOCAL hFile as PTR
     TRY
        hFile := FCreate("SomeFile.Txt")
     CATCH e as Exception
         // Handle Exception
     END TRY
     TRY
        FWriteLine(hFile, "sometext")
     CATCH e as Exception
         // Handle Exception
     END TRY
     TRY
        FClose(hFile)
     CATCH e as Exception
         // Handle Exception
     END TRY

```

**Comment by ecosSystem** [comment]
If you intend to find a global solution that fits all users that would really be the best way. For me it would even suffice, if the compiler accepts

recover using oErr as usual

This would allow a global replace in VS from

END SEQUENCE
to
RECOVER USING oErr as usual;IF oErr <> NIL;ErrorExec(oErr);ENDIF;END SEQUENCE

without the need to declare oErr as a local

I have written a small module which takes the transported output, performs some changes and copies the prg into the VS-solution. I could implement this replace easily. Before going live with our app I definitely will have many many cycles of transporting from VO. Internally we already work with our app, but we experience crashes which are hard to find without error-dialog..

**Comment by ecosSystem** [comment]
Robert,
If everybody else is happy with the current behaviour we should of course not change too much.
I (and I think Wolfgang too) would however be happy, if the compiler (through a switch "add missing recover-clauses" of course) could change
`Function Test(n as int) as void

BEGIN SEQUENCE
	if n = 0
		break
	endif
	ErrorTest(n)
END SEQUENCE

return
`
to
`Function Test(n as int) as void

BEGIN SEQUENCE
	if n = 0
		break
	endif
	ErrorTest(n)
RECOVER USING oErr as usual
	_SequenceRecover(oErr)
END SEQUENCE

return`

where _SequenceRecover() is an empty function in the runtime that does nothing.
We could then implement our own _SequenceRecover(oErr) function and call our errorhandler from there.
What do you think ?

**Comment by wriedmann** [comment]
I agree with Karl here, that would be the best option for VO migration, and should not hurt the performance of migrated applications.

**Comment by RobertvanderHulst** [comment]
Ok, to summarize your wishes:

- Leave the BEGIN SEQUENCE .. END SEQUENCE the way it is. So it will still be implemented with a try .. catch
- When the RECOVER USING is missing then generate a CATCH clause that calls a _SequenceRecover() like function in the runtime

So what about _REAL_ exceptions that happen in the block
- Should they also be sent to the _SequenceRecover() function ?
- Should they be caught and rethrown ?

**Comment by wriedmann** [comment]
IMHO they should be sent to the _SequenceRecover() function.
Documentation about that is important and should be added to the VO migration guide.

**Comment by ecosSystem** [comment]
If you catch and rethrow them then we don't need the extra recover. Both ways are ok to me. The extra recover has the advantage that I can still ignore errors when I want this. In cases where I have a recover now in VO I usually have installed a different errorblock which only calls _break(). If you generally catch exceptions that would again make problems.

**Comment by RobertvanderHulst** [comment]
Karl, Wolfgang,
Your requirements are not the same

The current implementation for a BEGIN SEQUENCE.. END SEQUENCE without RECOVER is this:

```
TRY
    TRY
       // the next line increments a counter in the runtime 
       // so the function _CanBreak() "knows" that a begin sequence .. end sequence is active
       CompilerServices.EnterBeginSequence()
       // the original code
    FINALLY
       // decrement the counter
       CompilerServices.ExitBeginSequence()
    END TRY
CATCH // no types, so all exceptions are caught
   NOP
END TRY
```  

And the line
`BREAK 42`
Is translated into
`THROW WrappedException{42}`

As you can see all exceptions are silently 'swallowed' by the CATCH block.
I suggest that we change that catch block to something like this

```
CATCH ex as Exception
   IF ex IS WrappedException VAR wrapped
       // The value from the BREAK statement gets stored in the Value property of a WrappedException object
        _SequenceRecover( wrapped:Value)    
   ELSE
        THROW // ex is not needed. This will make sure that the original callstack stays the same
   ENDIF         
END TRY
```

When there is a `RECOVER USING` clause then that code will be in the ex IS WrappedException block.
For example if looks like this
```
RECOVER USING e
      ? e
END SEQUENCE

```

Then the generated code in the CATCH block will become
```
CATCH ex as Exception
   IF ex IS WrappedException VAR wrapped
       // The value from the BREAK statement gets stored in the Value property of a WrappedException object
        e := wrapped:Value
        ? e
   ELSE
        THROW // ex is not needed. This will make sure that the original callstack stays the same
   ENDIF       
END TRY
```  
So to summarize my proposal:
- Always RE-THROW 'real' exceptions
- When no RECOVER USING is there, call a runtime function with the value stored inside the WrappedException

**Comment by wriedmann** [comment]
Hi Robert,
in VO, when there is no recover block, the error handler is called, and so the user is notified of a runtime error.
In X# currently there is no error message, and code code silently fails.
IMHO X# should possibily copy this behaviour, and it is easier for us to set a global error handler that is called when an error occurs, than change every place where a begin/end sequence without a recover is used.
But I leave it to you what the best way is.

**Comment by ecosSystem** [comment]
Robert, I think Wolfgang and I have the same requirement but perhaps different ways to explain it or find a solution. The main thing is that errors are shown and not swallowed. You certainly know best how to achieve this. Retry and Ignore are totally irrelevant to me, errors should be discovered and fixed. So if you think your solution is the best it will certainly be fine for me.

**Comment by ecosSystem** [comment]
Robert, I don't want to request too much of your time, but do you have any idea, how I can find the cause of the error below ? It cannot be reproduced, but seems to occur sometimes when closing a window. The error was caught by
`System.AppDomain.CurrentDomain:UnhandledException += UnhandledExceptionEventHandler{CurrentDomain_UnhandledException}`
Without the line-info it's very hard to locate the problem. I'm also very sure that there is no problem in DBFFENSTER:DISPATCH

BTW, you have a typo in German.resx in the runtime:
`<data name="USUALCONVERSIONERR" xml:space="preserve">
    <value>Konvertierungsfehler von USUAL ({0})  oin {1}</value>
  </data>`
this shoud be
`<data name="USUALCONVERSIONERR" xml:space="preserve">
    <value>Konvertierungsfehler von USUAL ({0}) in {1}</value>
  </data>`

Here is my error:

Fehlercode:                   	DATENTYP-FEHLER
Subsystem:                    	BASE
Argument:                     	USUAL
Typ:                          	PTR
Argument-Nummer:              	1
Beschreibung:                 	Konvertierungsfehler von USUAL (FLOAT)  oin PTR
Aufgerufen von:               	__USUAL:OP_IMPLICIT (Line: 0)
Aufgerufen von:               	DBFFENSTER:DISPATCH (Line: 0)
Aufgerufen von:               	MSKFENSTER:DISPATCH (Line: 0)
Aufgerufen von:               	__DOCAPP:DISPATCH (Line: 0)
Aufgerufen von:               	__WCDOCAPPWNDPROC (Line: 0)
Aufgerufen von:               	SENDMESSAGE (Line: 0)
Aufgerufen von:               	SHELLWINDOW:DISPATCH (Line: 0)
Aufgerufen von:               	MAINWINDOW:DISPATCH (Line: 0)
Aufgerufen von:               	MAIN:DISPATCH (Line: 0)
Aufgerufen von:               	__WCSHELLWNDPROC (Line: 0)
Aufgerufen von:               	TRANSLATEACCELERATOR (Line: 0)
Aufgerufen von:               	APP:EXEC (Line: 0)
Aufgerufen von:               	STARTPROGRAM (Line: 0)
Aufgerufen von:               	XAPPMAIN:START (Line: 0)
Aufgerufen von:               	START (Line: 0)

TIA Karl

**Comment by RobertvanderHulst** [comment]
Karl,
I agree that without line number this is difficult to find. But without source code it is even more difficult <g>.
But still I think I know what is going on.
The Dispatch() method is of Clipper calling convention. Its parameter is oEvent is therefore of type usual.
If you access the properties or fields  of this parameter you are therefore executing late bound code.
I guess you are assigning or casting one of the properties/fields of this event object to a pointer.
The value of this property is returned as a usual by the late bound code. Numeric values inside a usual are stored as LONG (when they are between -2^31 and +2^31) or as float.
The usual type allows casting from a LONG to a PTR. I think the value inside the usual is outside of the range that is valid for LONG values and therefore this error occurs.
To fix this:
- rename the parameter. FOr example from oEvent to uEvent
- add a local variable of type Event: LOCAL oEvent := uEvent AS Event
- you may have to prefix Event with @@ because Event is now also a keyword (I am not sure)
That should fix the problem because now there is no longer late bound code and you will directly cast a LONG or DWORD to a Pointer.
And thanks, I will fix the typo.

**Comment by ecosSystem** [comment]
Robert, thanks a lot, I will implement this immediately in DBFFENSTER and check all my other dispatch-methods

**Comment by RobertvanderHulst** [comment]
Karl,
I am sure this was the problem.
In general you should 'suspect' all places in your code where you are using _CAST.
_CAST means: compiler I know what I am doing.
And quite often that is not the case <g>.

**Comment by Karl1155** [comment]
Karl,

here´s a sample where such an "Conversion Error USUAL (FLOAT) to PTR" error is thrown. 

https://www.xsharp.eu/forum/german/1751-redimdatawindow#13088

This happens only if the Dispatch() event parameter is not typed.

Karl-Heinz

**Comment by ecosSystem** [comment]
Hi Robert,
will there be a solution for the BEGIN SEQUENCE / END SEQUENCE errorhandling-problem in 2.10 ? I did not see anything in the WhatsNew and also not in your ToDo-Lists...

Karl

**Comment by RobertvanderHulst** [comment]
Karl,
This is still on our todo list.
I will create a new ticket that describes the steps that we'll take

**Comment by RobertvanderHulst** [comment]
See https://github.com/X-Sharp/XSharpPublic/issues/916

**Comment by ecosSystem** [comment]
Robert,
as Chris stated, my main problem is errors within BEGIN SEQUENCE / END SEQUENCE without recover. You have opened #916 yourself which is still open. So no need to add a new ticket I think.

**Comment by cpyrgas** [comment]
Karl, please test the new /vo17+ compiler/project option that is implemented in the next (2.12) build and let us know if it works well for you. Enabling this should make the runtime by default display the error window for errors happening inside BEGIN...END SEQUENCE. Pressing "Ignore" will move program execution right after the END SEQUENCE statement (still not possible to resume to the line just after the error)

**Comment by cpyrgas** [comment]
Note for future reference: Related to https://github.com/X-Sharp/XSharpPublic/issues/111 and https://github.com/X-Sharp/XSharpPublic/issues/916

**Comment by ecosSystem** [comment]
Chris, Robert,
can you give me access to the beta of 2.12 ? I could test it a little bit this week. Next week is a short one in Germany and the week after I'll be on holiday...

**Comment by ecosSystem** [comment]
Robert,
I've just installed 2.12 and tried the switch "Compatible BEGIN SEQUENCE .. END SEQUENCE"
It does not work for me.

`FUNCTION ErrorTest() AS VOID
LOCAL x AS STRING
x := &("1/1")
return
`
creates the expected error


`FUNCTION ErrorTest() AS VOID
LOCAL x AS STRING

BEGIN SEQUENCE
x := &("1/1")
END SEQUENCE

RETURN
`
still does nothing. Also the function _SequenceRecover is never called.
Did I miss any other switch ?

Karl

**Comment by ecosSystem** [comment]
"Compatible BEGIN SEQUENCE..END SEQUENCE" checkbox...

**Comment by cpyrgas** [comment]
Please read my comment just above yours :)

**Comment by ecosSystem** [comment]
I only see your question. I enabled the checkbox in the Dialect page. Where can I set /vo17 ?

**Comment by cpyrgas** [comment]
Ah, I am sorry you are right...This is indeed the correct checkbox, but seems VS ignores it! For now, please go to the "Build" page and add "/vo17+" (without the quotes) in the Extra Command line options.

Robert, can you have a look into this? @RobertvanderHulst

**Comment by ecosSystem** [comment]
This seems to work at first sight, I'll do further tests today...

**Comment by RobertvanderHulst** [comment]
There is a problem in the build support code.
Download https://1drv.ms/u/s!An3aSY_HolAD4h6A_A5jO7MHzIvP?e=EMsBSn and copy the contents to the "c:\Program Files (x86)\XSharp\MsBuild"  folder and restart VS. That should fix it.

**Comment by ecosSystem** [comment]
Now both the setting and the errorhandling works! Thanks a lot, I let you know if I have any problems with it...
